### PR TITLE
Adding a data persistence level

### PR DESCRIPTION
**Pull request by IuriiSokolov**
Completing the module 3 - Persistence Level (SQL, NoSQL) + DAL
Adding EF core models, DbContext and repositories for each table with CRUD

### CODE REVIEW

File: TicketingSystem.ApiService/Program.cs [hunk 1]
```diff
@@ -1,43 +1,21 @@
-using TicketingSystem.Common;
+using TicketingSystem.ApiService.Extensions;
+using TicketingSystem.Common.Context;
```
Comment (olegbaslak) [inline TicketingSystem.ApiService/Program.cs:3]: Just a small suggestion here. I would structure the solution by separating different layers into different C# projects from the start. You can do it in a separate PR.

File: TicketingSystem.ApiService/Extensions/ServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿using TicketingSystem.ApiService.Repositories.EventRepository;
+using TicketingSystem.ApiService.Repositories.PersonRepository;
+using TicketingSystem.ApiService.Repositories.VenueRepository;
+using TicketingSystem.ApiService.Repositories.SeatRepository;
+using TicketingSystem.ApiService.Repositories.TickerRepository;
+
+namespace TicketingSystem.ApiService.Extensions
```
Comment (olegbaslak) [inline TicketingSystem.ApiService/Extensions/ServiceCollectionExtensions.cs]: Since you may have a lot of extensions, I'd name the folder `DependencyInjection` instead of `Extensions`. That folder only contains service collection extensions related to service registration. Each C# project added a separate module may have it's own `DependencyInjection` folder and extensions.

File: TicketingSystem.ApiService/Program.cs (hunk 1 above)
Comment (olegbaslak) [reply TicketingSystem.ApiService/Program.cs:3]: You'll notice that you'll have to separate Repository Implementation and Interface, so DAL layer only contains implementation and application (services) layer contains the interface

File: TicketingSystem.ApiService/Repositories/PaymentRepository/PaymentRepository.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
+﻿using Microsoft.EntityFrameworkCore;
+using TicketingSystem.Common.Context;
+using TicketingSystem.Common.Model.Database;
+
+namespace TicketingSystem.ApiService.Repositories.PaymentRepository
+{
+    public class PaymentRepository : IPaymentRepository
+    {
+        TicketingDbContext _context;
```
Comment (olegbaslak) [inline TicketingSystem.ApiService/Repositories/PaymentRepository/PaymentRepository.cs]: Just a rule for sake of clarity: all injected services should be `private readonly`. This is kind of de-facto standard :D

File: TicketingSystem.Common/Context/ITicketingDbContext.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using TicketingSystem.Common.Model.Database;
+
+namespace TicketingSystem.Common.Context
+{
+    public interface ITicketingDbContext
```
Comment (olegbaslak) [inline TicketingSystem.Common/Context/ITicketingDbContext.cs]: How do you use this interface?

File: TicketingSystem.Common/Context/TicketingDbContext.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace TicketingSystem.Common.Context
+{
+    public class TicketingDbContext(DbContextOptions<TicketingDbContext> options) : DbContext(options), ITicketingDbContext
+    {
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            base.OnModelCreating(modelBuilder);
+
+            modelBuilder.Entity<Ticket>()
```
Comment (olegbaslak) [inline TicketingSystem.Common/Context/TicketingDbContext.cs]: In middle/big projects you will fast notice that db context class becomes a mess if you put everything here. For that case in EF Core there's a possibility to define configuration for each Entity and apply it here in `OnModelCreating`.

Use `TicketEntityConfiguration : IEntityTypeConfiguration<Ticket>` and apply it via `modelBuilder.ApplyConfiguration(new TicketEntityConfiguration())`. Inside of `TicketEntityConfiguration ` you can define configuration for each column and its relations, table name, keys, etc. For example your `SeatStatusRow` entity will get rid of EF Core attributes on class and properties if you extract configurations into `IEntityTypeConfiguration` objects.

Comment (olegbaslak) [reply TicketingSystem.Common/Context/TicketingDbContext.cs]: Initial values may still be there for sake of testing, but from my point of view, I would just separate data seeding into a separate object and run it during startup (in `Program.cs`). Of course you wouldn't do that in prod, but for the course project it's totally okay.

File: TicketingSystem.Common/Model/Database/Seat.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+﻿using TicketingSystem.Common.Model.Database.Enums;
+
+namespace TicketingSystem.Common.Model.Database
+{
+    public class Seat
+    {
+        public int SeatId { get; set; }
```
Comment (olegbaslak) [inline TicketingSystem.Common/Model/Database/Seat.cs]: Your DB schema missing `Section` and `Row` to identify specific seats. You'll need them later to return sections for a venue and specific section seats for events.

Also, ticket admission type is missing:
> The are two seat types - specific designated place in a row and general admission. General admission means patrons can choose to sit anywhere within a particular section (for example, the dance floor)/

File: TicketingSystem.Common/Model/Database/Enums/SeatStatus.cs [hunk 7]
```diff
@@ -0,0 +1,9 @@
+﻿namespace TicketingSystem.Common.Model.Database.Enums
+{
+    public enum SeatStatus
+    {
+        Free = 1,
```
Comment (olegbaslak) [inline TicketingSystem.Common/Model/Database/Enums/SeatStatus.cs]: Default value for enum is `0`, so it's not a good idea to start with `1`

File: TicketingSystem.Common/Model/Database/Seat.cs (hunk 6 above)
Comment (olegbaslak) [reply TicketingSystem.Common/Model/Database/Seat.cs]: Also I do not see `Prices` table:

> **Prices –** Information about the ticket price levels (Adult, Child, VIP, etc.)

File: TicketingSystem.Common/Model/Database/Payment.cs [hunk 8]
```diff
@@ -0,0 +1,10 @@
+﻿namespace TicketingSystem.Common.Model.Database
+{
+    public class Payment
+    {
+        public int PaymentId { get; set; }
+        public required DateTime PaymentTime { get; set; }
+
+        public Ticket Ticket { get; set; } = null!;
```
Comment (olegbaslak) [inline TicketingSystem.Common/Model/Database/Payment.cs]: Payment may contain more than one ticket

**Review by olegbaslak — CHANGES_REQUESTED** [review]

File: TicketingSystem.ApiService/Program.cs (hunk 1 above)
Comment (IuriiSokolov) [reply TicketingSystem.ApiService/Program.cs:3]: Will do in the following PRs

File: TicketingSystem.ApiService/Extensions/ServiceCollectionExtensions.cs (hunk 2 above)
Comment (IuriiSokolov) [reply TicketingSystem.ApiService/Extensions/ServiceCollectionExtensions.cs]: Renamed

File: TicketingSystem.ApiService/Repositories/PaymentRepository/PaymentRepository.cs (hunk 3 above)
Comment (IuriiSokolov) [reply TicketingSystem.ApiService/Repositories/PaymentRepository/PaymentRepository.cs]: Added private readonly to all context usages

File: TicketingSystem.Common/Context/ITicketingDbContext.cs (hunk 4 above)
Comment (IuriiSokolov) [reply TicketingSystem.Common/Context/ITicketingDbContext.cs]: Interface is removed

File: TicketingSystem.Common/Context/TicketingDbContext.cs (hunk 5 above)
Comment (IuriiSokolov) [reply TicketingSystem.Common/Context/TicketingDbContext.cs]: Added configurations for all entities
I only let efcore configure the many-to-many relationship between Carts and Tickets by convention because when I tried to do it in config the table CartsTickets didn't create

Verified that the migration is the same with or without configs

File: TicketingSystem.Common/Model/Database/Seat.cs (hunk 6 above)
Comment (IuriiSokolov) [reply TicketingSystem.Common/Model/Database/Seat.cs]: Added tables Sections, PriceCategories, SeatTypes
Added int RowNumber to Seats

File: TicketingSystem.Common/Model/Database/Enums/SeatStatus.cs (hunk 7 above)
Comment (IuriiSokolov) [reply TicketingSystem.Common/Model/Database/Enums/SeatStatus.cs]: Changed it to 0

File: TicketingSystem.Common/Model/Database/Payment.cs (hunk 8 above)
Comment (IuriiSokolov) [reply TicketingSystem.Common/Model/Database/Payment.cs]: Payments are now linked to Carts

**Review by olegbaslak — APPROVED** [review]
